plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucketServerPullRequest.ts

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { InputError } from '@backstage/errors';
+import {
+  getBitbucketServerRequestOptions,
+  ScmIntegrationRegistry,
+} from '@backstage/integration';
+import { createTemplateAction } from '@backstage/plugin-scaffolder-node';
+import fetch, { RequestInit, Response } from 'node-fetch';
+import { parseRepoUrl } from './util';
+import { Config } from '@backstage/config';
+
+const createPullRequest = async (opts: {
+  project: string;
+  repo: string;
+  title: string;
+  description?: string;
+  toRef: object;
+  fromRef: object;
+  authorization: string;
+  apiBaseUrl: string;
+}) => {
+  const {
+    project,
+    repo,
+    title,
+    description,
+    toRef,
+    fromRef,
+    authorization,
+    apiBaseUrl,
+  } = opts;
+
+  let response: Response;
+  const data: RequestInit = {
+    method: 'POST',
+    body: JSON.stringify({
+      title: title,
+      description: description,
+      state: 'OPEN',
+      open: true,
+      closed: false,
+      locked: true,
+      toRef: toRef,
+      fromRef: fromRef,
+    }),
+    headers: {
+      Authorization: authorization,
+      'Content-Type': 'application/json',
+    },
+  };
+
+  try {
+    response = await fetch(
+      `${apiBaseUrl}/projects/${project}/repos/${repo}/pull-requests`,
+      data,
+    );
+  } catch (e) {
+    throw new Error(`Unable to create pull-reqeusts, ${e}`);
+  }
+
+  if (response.status !== 201) {
+    throw new Error(
+      `Unable to create pull requests, ${response.status} ${
+        response.statusText
+      }, ${await response.text()}`,
+    );
+  }
+
+  const r = await response.json();
+  return `${r.links.self[0].href}`;
+};
+const findBranches = async (opts: {
+  project: string;
+  repo: string;
+  branchName: string;
+  authorization: string;
+  apiBaseUrl: string;
+}) => {
+  const { project, repo, branchName, authorization, apiBaseUrl } = opts;
+
+  let response: Response;
+  const options: RequestInit = {
+    method: 'GET',
+    headers: {
+      Authorization: authorization,
+      'Content-Type': 'application/json',
+    },
+  };
+
+  try {
+    response = await fetch(
+      `${apiBaseUrl}/projects/${project}/repos/${repo}/branches?boostMatches=true&filterText=${branchName}`,
```

Comment:
possibly a bit unfortunate to use a list/search endpoint for finding the details of a branch - feels like there's gotta be a more exact one to use which directly targets the branch by name :) but i'm not familiar enough with their API to know for sure

Reply:
I also think so. However, this was the only API I found that accurately responded with the fromRefo and toRefo information for the pull request.

Reply:
I just realized: All of the parts of the URL should be `encodeURIComponent`-ed

Reply:
changed it.